src/components/tabController/TabBar.tsx

```diff
@@ -128,364 +128,289 @@ interface Props extends TabControllerBarProps, BaseComponentInjectedProps, Forwa
-    return contextItems || propsItems;
-  }
-
-  getItemsCount() {
-    if (this.items) {
-      return _.size(this.items);
-    } else {
-      return React.Children.count(this.children);
-    }
-  }
-
-  getSnapBreakpoints() {
-    const {centerSelected} = this.props;
-    const {itemsWidths, itemsOffsets} = this.state;
-
-    if (itemsWidths.length > 0 && itemsOffsets.length > 0 && centerSelected) {
-      return _.times(itemsWidths.length, index => {
-        const screenCenter = this.containerWidth / 2;
-        const itemOffset = itemsOffsets[index];
-        const itemWidth = itemsWidths[index];
-        return itemOffset - screenCenter + itemWidth / 2;
-      });
-    }
-  }
-
-  registerTabItems() {
-    const {registerTabItems} = this.context;
-    const {items} = this.props;
+const TabBar = (props: Props) => {
+  const {
+    items: propsItems,
+    height,
+    enableShadow,
+    shadowStyle: propsShadowStyle,
+    // minTabsForScroll,
+    indicatorStyle,
+    labelStyle,
+    selectedLabelStyle,
+    labelColor,
+    selectedLabelColor,
+    uppercase,
+    iconColor,
+    selectedIconColor,
+    activeBackgroundColor,
+    backgroundColor,
+    containerWidth: propsContainerWidth,
+    centerSelected: propsCenterSelected,
+    containerStyle,
+    testID,
+    children: propsChildren
+  } = props;
+
+  const context = useContext(TabBarContext);
+  // @ts-ignore // TODO: typescript
+  const {itemStates, items: contextItems, selectedIndex, currentPage, targetPage, registerTabItems} = context;
+
+  const children = useRef<Props['children']>(_.filter(propsChildren, (child: ChildProps) => !!child));
+
+  const _registerTabItems = () => {
     const ignoredItems: number[] = [];
     let itemsCount;
 
-    if (items) {
-      itemsCount = _.size(items);
-      _.forEach(items, (item, index) => {
+    if (propsItems) {
+      itemsCount = _.size(propsItems);
+      _.forEach(propsItems, (item, index) => {
         if (item.ignore) {
           ignoredItems.push(index);
         }
       });
       // TODO: deprecate with props.children
     } else {
-      itemsCount = React.Children.count(this.children);
+      itemsCount = React.Children.count(children.current);
       // @ts-ignore TODO: typescript - not sure if this can be solved
-      React.Children.toArray(this.children).forEach((child: ChildProps, index: number) => {
+      React.Children.toArray(children.current).forEach((child: ChildProps, index: number) => {
         if (child.props.ignore) {
           ignoredItems.push(index);
         }
       });
     }
 
     registerTabItems(itemsCount, ignoredItems);
-  }
+  };
+
+  useEffect(() => {
+    if (propsChildren) {
+      LogService.warn('uilib: Please pass the "items" prop to TabController.TabBar instead of children');
+    }
+
+    if ((propsItems || children.current) && !contextItems) {
+      _registerTabItems();
+    }
+  }, []);
+
+  const [itemsWidths, setItemsWidths] = useState<number[]>([]);
```

Comment:
consider moving some parts of the code to a separate custom hook (it does't have to be every part)
This will cleanup some of TabBar's code and create better separation. 
For instance, maybe we can move the whole logic that calculates itemWidth and itemOffset to a separate hook.

Reply:
This is done in the second PR ✅